### PR TITLE
First pass on backwards compatibility principle

### LINKED ISSUE #174: add principle on backwards compatibility / supporting existing content (HTML Design Principle 2.1)
The HTML design principles has a section [2.1. Support Existing Content](https://www.w3.org/TR/html-design-principles/#support-existing-content) about continuing to support existing content.  As a followup to https://github.com/w3ctag/design-principles/issues/126#issuecomment-610117172 @fantasai's blog post has a [different take](http://fantasai.inkedblade.net/weblog/2019/designing-css/#compat) on both this and forwards compatibility (#175).

### PR DESCRIPTION
**Pull request by LeaVerou**
First pass, let's iterate. 
Resolves #174

***
<a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/pull/354.html" title="Last updated on Jan 24, 2024, 10:18 AM UTC (3b26d40)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/354/3830de1...3b26d40.html" title="Last updated on Jan 24, 2024, 10:18 AM UTC (3b26d40)">Diff</a>

### CODE REVIEW

**Comment by torgo** [comment]
As discussed in today's TAG call maybe we need to amend this to reference the evergreen web finding.

File: index.bs [hunk 1]
```diff
@@ -263,6 +263,30 @@ so that it works across screen sizes,
 and allows users to choose their own font and font size without causing text to overflow.
 </div>
 
+<h3 id="backwards-compat">Support Existing Content</h3>
+
+Existing content often relies upon expected user agent processing and behavior to function as intended.
+Processing requirements should be specified to ensure that user agents implementing this specification will be able to handle most existing content.
```
Comment (torgo) [inline index.bs:269]:
```suggestion
Processing requirements should be specified to ensure that user agents will be able to handle most existing content.
```

File: index.bs [hunk 2]
```diff
@@ -263,6 +263,30 @@ so that it works across screen sizes,
...
 
+<h3 id="backwards-compat">Support Existing Content</h3>
+
+Existing content often relies upon expected user agent processing and behavior to function as intended.
+Processing requirements should be specified to ensure that user agents will be able to handle most existing content.
+In particular, it should be possible to process existing web content in current user agents and get results that are compatible with the existing expectations of users and authors,
+based on the behavior of existing browsers.
+
+Content relying on existing browser behavior can take many forms.
+It may rely on elements, attributes or APIs that are part of earlier specifications, or on features that are entirely proprietary.
```
Comment (kenchris) [inline index.bs:274]: Make these a list instead?

File: index.bs [hunk 3]
```diff
@@ -263,6 +263,30 @@ so that it works across screen sizes,
...
+It may depend on specific error handling rules. It may depend on certain method names not being defined.
```
Comment (kenchris) [inline index.bs:275]: Do we have any example of that?

File: index.bs [hunk 4]
```diff
@@ -263,6 +263,30 @@ so that it works across screen sizes,
...
+In rare cases, it may even depend on a feature from earlier specifications not being implemented as specified.
```
Comment (kenchris) [inline index.bs:276]: In which case we should probably change the specifications

File: index.bs [hunk 5]
```diff
@@ -263,6 +263,30 @@ so that it works across screen sizes,
...
+
+When considering changes to legacy features or behavior, relative to current implementations and author expectations, the following questions should be considered:
+
+* Does a significant quantity of existing content depend on the feature or behavior?
```
Comment (kenchris) [inline index.bs:280]: Does it have to be significant quantity? It could be a feature that is rarely used but it would break quite important services, or it might even break things inside intranets. I think showModalDialog was one such case

File: index.bs [hunk 6]
```diff
@@ -263,6 +263,30 @@ so that it works across screen sizes,
...
+* Does any of the dependent content occur on particularly popular websites?
```
Comment (kenchris) [inline index.bs:281]: I am not a big fan or talking about popularity. An intranet site at a large might be just as important. It might also be used on a very unpopular website that everyone uses, like for declaring taxes ;-)

maybe "high volume websites or intranet sites"

We have made such changes before in such cases, what really mattered was making a public plan and giving people time to adjust existing content

File: index.bs [hunk 7]
```diff
@@ -263,6 +263,30 @@ so that it works across screen sizes,
...
+* Is the dependent content genuinely intended for consumption, rather than occurring solely in test cases or examples?
+* Is the dependent content on the public web, rather than found solely on internal sites with a controlled user environment?
+* Does the dependent content currently work as intended in multiple popular user agents, rather than explicitly targeting only one particular user agent?
+
+The benefit of the proposed change should be weighed against the likely cost of breaking content, as measured by these criteria.
+In some cases, it may be desirable to make a nonstandard feature or behavior part of the web platform, if it satisfies a valid use case.
+However, the fact that something is part of the web platform does not by itself mean that relying on it is condoned or encouraged.
```
Comment (kenchris) [inline index.bs:288]: Yes sometimes it doesn't make sense to make that change, but other times it does, what matters is making a public plan for the change with a roadmap and time line so that people and companies can prepare and adjust

**Comment by torgo** [comment]
We discussed in the breakout call today and agreed this needs some additional rewording and needs to be condensed down  - also it needs to be adapted to align with the [style guide](https://github.com/w3ctag/process/blob/master/style-guide.md).

File: index.bs [hunk 8]
```diff
@@ -263,6 +263,30 @@ so that it works across screen sizes,
...
+When considering changes to legacy features or behavior, relative to current implementations and author expectations, the following questions should be considered:
+
+* Does a significant quantity of existing content depend on the feature or behavior?
+* Does any of the dependent content occur on particularly popular websites?
+* Is the dependent content genuinely intended for consumption, rather than occurring solely in test cases or examples?
+* Is the dependent content on the public web, rather than found solely on internal sites with a controlled user environment?
+* Does the dependent content currently work as intended in multiple popular user agents, rather than explicitly targeting only one particular user agent?
+
+The benefit of the proposed change should be weighed against the likely cost of breaking content, as measured by these criteria.
+In some cases, it may be desirable to make a nonstandard feature or behavior part of the web platform, if it satisfies a valid use case.
```
Comment (maxpassion) [inline index.bs:287]: Not sure whether we need to emphasize this nonstandard feature case?

File: index.bs (hunk 4 above)
Comment (atanassov) [inline index.bs:276]:
```suggestion
Content relying on existing browser behavior can take many forms, such as.

- Rely on elements, attributes or APIs that are part of earlier specifications, or on features that are entirely proprietary.
- Depend on specific error handling rules.
- Depend on certain method names not being defined.
- Depend on a feature from earlier specifications not being implemented as specified.
```

File: index.bs (hunk 3 above)
Comment (atanassov) [reply index.bs:275]: Lots of example is UA sniffing content. Not saying this is a good example but the tail end of the web still depends on it.

File: index.bs (hunk 5 above)
Comment (torgo) [reply index.bs:282]:
```suggestion
* Would removal of this feature or behavior have significant impact on existing content?
```

File: index.bs [hunk 9]
```diff
@@ -263,6 +263,32 @@ so that it works across screen sizes,
...
+However, the fact that something is part of the web platform does not by itself mean that relying on it is condoned or encouraged.
+
```
Comment (LeaVerou) [inline index.bs:291]:
```suggestion

<div class="example">

Many sites use broken markup, such as badly nested elements (`<b>a<i>b</b>c</i>`), and both authors and users have expectations based on the error handling used by legacy user agents. We need to define processing requirements that remain compatible with the expected handling of such content.

Some sites rely on the `<u>` element giving the presentational effect of an underline.
</div>
```

File: index.bs (hunk 6 above)
Comment (torgo) [reply index.bs:283]:
```suggestion
* Does any of the dependent content occur on particularly high volume websites?
```

File: index.bs [hunk 10]
```diff
@@ -263,6 +263,39 @@ so that it works across screen sizes,
...
+When considering changes to legacy features or behavior, relative to current implementations and author expectations, the following questions should be considered:
+
+* Does a significant quantity of existing content depend on the feature or behavior?
+* Does any of the dependent content occur on particularly popular websites?
+* Is the dependent content genuinely intended for consumption, rather than occurring solely in test cases or examples?
+* Is the dependent content on the public web, rather than found solely on internal sites with a controlled user environment?
+* Does the dependent content currently work as intended in multiple popular user agents, rather than explicitly targeting only one particular user agent?
+
+The benefit of the proposed change should be weighed against the likely cost of breaking content, as measured by these criteria.
+In some cases, it may be desirable to make a nonstandard feature or behavior part of the web platform, if it satisfies a valid use case.
```
Comment (LeaVerou) [inline index.bs:289]:
```suggestion
```

File: index.bs [hunk 11]
```diff
@@ -263,6 +263,39 @@ so that it works across screen sizes,
...
+However, the fact that something is part of the web platform does not by itself mean that relying on it is condoned or encouraged.
```
Comment (LeaVerou) [inline index.bs:290]:
```suggestion
```

File: index.bs (hunk 10 above)
Comment (torgo) [inline index.bs:289]:
```suggestion
```

File: index.bs [hunk 12]
```diff
@@ -263,6 +263,37 @@ so that it works across screen sizes,
 and allows users to choose their own font and font size without causing text to overflow.
 </div>
 
+<h3 id="backwards-compat">Support Existing Content</h3>
+
+Existing content often relies upon expected user agent processing and behavior to function as intended.
+Processing requirements should be specified to ensure that user agents will be able to handle most existing content.
+In particular, it should be possible to process existing web content in current user agents and get results that are compatible with the existing expectations of users and authors,
+based on the behavior of existing browsers.
```
Comment (torgo) [inline index.bs:271]:
```suggestion
based on the behavior of existing user agents.
```

File: index.bs [hunk 13]
```diff
@@ -263,6 +263,37 @@ so that it works across screen sizes,
...
+
+Content relying on existing browser behavior can take many forms, such as:
```
Comment (torgo) [inline index.bs:273]:
```suggestion
Content relying on existing user agent behavior can take many forms, such as:
```

File: index.bs [hunk 14]
```diff
@@ -263,6 +263,37 @@ so that it works across screen sizes,
...
+
+- Relying on elements, attributes or APIs that are part of earlier specifications, or on features that are entirely proprietary.
```
Comment (torgo) [inline index.bs:275]:
```suggestion
- Relying features that are entirely proprietary.
```

Comment (torgo) [reply index.bs:275]: Removes redundancies and keeps things short.

File: index.bs (hunk 13 above)
Comment (torgo) [reply index.bs:273]: consistency

File: index.bs (hunk 12 above)
Comment (torgo) [reply index.bs:271]: consistency

File: index.bs [hunk 15]
```diff
@@ -263,6 +263,37 @@ so that it works across screen sizes,
...
+- Depending on specific error handling rules.
+- Depending on certain method names not being defined.
+- Depending on a feature from earlier specifications not being implemented as specified.
+
+When considering changes to legacy features or behavior, relative to current implementations and author expectations, the following questions should be considered:
```
Comment (torgo) [inline index.bs:280]:
```suggestion
When considering changes to legacy features, relative to current implementations and author expectations, the following questions should be considered:
```

File: index.bs [hunk 16]
```diff
@@ -263,6 +263,37 @@ so that it works across screen sizes,
...
+
+* Would removal of this feature or behavior have significant impact on existing content?
```
Comment (torgo) [inline index.bs:282]:
```suggestion
* Would removal of this feature have significant impact on existing content?
```

File: index.bs (hunk 15 above)
Comment (torgo) [inline index.bs:280]:
```suggestion
The following questions should be answered when considering removal of a feature that would have significant impact on existing content:
```

File: index.bs [hunk 17]
```diff
@@ -263,6 +263,37 @@ so that it works across screen sizes,
...
+* Does any of the dependent content occur on particularly high volume websites?
+* Is the dependent content genuinely intended for consumption, rather than occurring solely in test cases or examples?
+* Is the dependent content on the public web, rather than found solely on internal sites with a controlled user environment?
+* Does the dependent content currently work as intended in multiple popular user agents, rather than explicitly targeting only one particular user agent?
+
+The benefit of the proposed change should be weighed against the likely cost of breaking content, as measured by these criteria.
+
+<div class="example">
+
+Many sites use broken markup, such as badly nested elements (`<b>a<i>b</b>c</i>`), and both authors and users have expectations based on the error handling used by legacy user agents. We need to define processing requirements that remain compatible with the expected handling of such content.
```
Comment (torgo) [inline index.bs:292]:
```suggestion
Many sites use broken markup, such as badly nested elements (`<b>a<i>b</b>c</i>`), and authors have expectations based on the existing error handling results of legacy user agents that such broken markup will still work. We need to define processing requirements that remain compatible with the expected handling of such content.
```

File: index.bs [hunk 18]
```diff
@@ -263,6 +263,37 @@ so that it works across screen sizes,
...
+
+Some sites rely on the `<u>` element giving the presentational effect of an underline.
+
```
Comment (torgo) [inline index.bs:295]:
```suggestion
```

File: index.bs [hunk 19]
```diff
@@ -263,6 +263,37 @@ so that it works across screen sizes,
 and allows users to choose their own font and font size without causing text to overflow.
 </div>
 
+<h3 id="backwards-compat">Support Existing Content</h3>
+
```
Comment (torgo) [inline index.bs:267]: We need this to start with actionable advice, per the styleguide.

File: index.bs (hunk 16 above)
[the remark this answers is not in the document]
Comment (LeaVerou) [reply index.bs:282]: Why?

File: index.bs [hunk 20]
```diff
@@ -263,6 +263,34 @@ so that it works across screen sizes,
...
+* Does any of the dependent content occur on particularly high volume websites?
+* Is the dependent content genuinely intended for consumption, rather than occurring solely in test cases or examples?
+* Is the dependent content on the public web, rather than found solely on internal sites with a controlled user environment?
+* Does the dependent content currently work as intended in multiple popular user agents, rather than explicitly targeting only one particular user agent?
+
+The benefit of the proposed change should be weighed against the likely cost of breaking content, as measured by these criteria.
+
+<div class="example">
+
+Many sites use broken markup, such as badly nested elements (`<b>a<i>b</b>c</i>`), and authors have expectations based on the existing error handling results of legacy user agents that such broken markup will still work. We need to define processing requirements that remain compatible with the expected handling of such content.
```
Comment (dbaron) [inline index.bs:292]: "legacy" here also seems to have negative meaning (implying "outdated"), whereas I think what's intended here may be something closer to "existing"

File: index.bs [hunk 21]
```diff
@@ -263,6 +263,34 @@ so that it works across screen sizes,
...
+based on the behavior of existing user agents.
+
+Content relying on existing user agent behavior can take many forms, such as:
+
+- Relying on features that are entirely proprietary.
+- Depending on specific error handling rules.
+- Depending on certain method names not being defined.
+- Depending on a feature from earlier specifications not being implemented as specified.
+
+When considering changes to legacy features, relative to current implementations and author expectations, the following questions should be considered:
```
Comment (dbaron) [inline index.bs:280]: I think another question worth considering is something like "is the feature specified well enough and open enough that implementations that don't already implement it can do so".  (If the answer is no, I think that's an argument in favor of either (a) removal or (b) doing the work to change the answer to yes.)

File: index.bs [hunk 22]
```diff
@@ -263,6 +263,34 @@ so that it works across screen sizes,
 and allows users to choose their own font and font size without causing text to overflow.
 </div>
 
+<h3 id="backwards-compat">Support Existing Content</h3>
+
+Existing content often relies upon expected user agent processing and behavior to function as intended.
```
Comment (torgo) [inline index.bs:268]:
```suggestion
When making changes to web APIs, consider the impact on existing web content.
Existing content often relies upon expected user agent processing and behavior to function as intended.
```

File: index.bs (hunk 21 above)
[the remark this answers is not in the document]
Comment (torgo) [reply index.bs:280]:
```suggestion
When considering changes to already widely implemented features, relative to current implementations and author expectations, the following questions should be considered:
```

File: index.bs [hunk 23]
```diff
@@ -263,6 +263,34 @@ so that it works across screen sizes,
...
+Processing requirements should be specified to ensure that user agents will be able to handle most existing content.
+In particular, it should be possible to process existing web content in current user agents and get results that are compatible with the existing expectations of users and authors,
+based on the behavior of existing user agents.
+
+Content relying on existing user agent behavior can take many forms, such as:
+
+- Relying on features that are entirely proprietary.
```
Comment (plinss) [inline index.bs:275]: Perhaps an additional bullet point here describing specified and widely implemented features. The rest mostly talks about un(der)specified behavior, and it seems like we're missing the main thing.

File: index.bs [hunk 24]
```diff
@@ -263,6 +263,34 @@ so that it works across screen sizes,
...
+- Depending on specific error handling rules.
+- Depending on certain method names not being defined.
+- Depending on a feature from earlier specifications not being implemented as specified.
+
+When considering changes to legacy features, relative to current implementations and author expectations, the following questions should be considered:
+
+* Would removal of this feature or behavior have significant impact on existing content?
```
Comment (plinss) [inline index.bs:282]: This could also talk about modifying existing behavior

File: index.bs (hunk 21 above)
[the remark this answers is not in the document]
Comment (LeaVerou) [reply index.bs:281]: Not sure either is what we need. It's about widely *used* features, which are not necessarily widely implemented (e.g. devs may use a feature that is not widely implemented if it does not hamper the user experience in non supporting UAs), and are not necessarily legacy either.

```suggestion
When considering changes to widely used features, relative to current implementations and author expectations, the following questions should be considered:
```

File: index.bs [hunk 25]
```diff
@@ -296,7 +325,7 @@ commonality and complexity are not always correlated.
 <div class=example>
 Sanitizing HTML to prevent XSS attacks is a complex process
 that requires extensive security knowledge,
-however the [Sanitizer API](https://wicg.github.io/sanitizer-api/) provides a shortcut for this common use case. 
+however the [Sanitizer API](https://wicg.github.io/sanitizer-api/) provides a shortcut for this common use case.
```
Comment (LeaVerou) [inline index.bs:328]: This change seems to be unrelated to this PR.

File: index.bs [hunk 26]
```diff
@@ -277,7 +306,7 @@ although they may be harder to find.
 Simpler features are easier for user agents to implement and test,
 more likely to be interoperable,
 and easier for authors to understand.
-It is especially important to design your feature so that 
+It is especially important to design your feature so that
```
Comment (LeaVerou) [inline index.bs:309]: unrelated change

File: index.bs [hunk 27]
```diff
@@ -263,6 +263,35 @@ so that it works across screen sizes,
 and allows users to choose their own font and font size without causing text to overflow.
 </div>
 
+<h3 id="backwards-compat">Support Existing Content</h3>
```
Comment (LeaVerou) [inline index.bs:266]:
```suggestion
<h3 id="backwards-compat">New features should not break existing content</h3>
```

File: index.bs [hunk 28]
```diff
@@ -263,6 +263,35 @@ so that it works across screen sizes,
 and allows users to choose their own font and font size without causing text to overflow.
 </div>
 
+<h3 id="backwards-compat">New features should not break existing content</h3>
+
+When making changes to web APIs, consider the impact on existing web content.
+Existing content often relies upon expected user agent processing and behavior to function as intended.
```
Comment (martinthomson) [inline index.bs:269]:
```suggestion
Existing content often relies on existing user agent behavior.
```

File: index.bs [hunk 29]
```diff
@@ -263,6 +263,35 @@ so that it works across screen sizes,
...
+Processing requirements should be specified to ensure that user agents will be able to handle most existing content.
```
Comment (martinthomson) [inline index.bs:270]:
```suggestion

```

Why say this at all?

File: index.bs [hunk 30]
```diff
@@ -263,6 +263,35 @@ so that it works across screen sizes,
...
+In particular, it should be possible to process existing web content in current user agents and get results that are compatible with the existing expectations of users and authors,
+based on the behavior of existing user agents.
```
Comment (martinthomson) [inline index.bs:272]:
```suggestion
It should be possible to process existing web content in current user agents and
get results that are compatible with preexisting expectations of users and authors.
This will generally match the behavior of user agents prior to the introduction of the new feature.
```

File: index.bs [hunk 31]
```diff
@@ -263,6 +263,35 @@ so that it works across screen sizes,
...
+
+Content relying on existing user agent behavior can take many forms, such as:
```
Comment (martinthomson) [inline index.bs:274]:
```suggestion
There are cases where the behavior that content depends upon does not match what is documented in standards.
This includes:
```

File: index.bs [hunk 32]
```diff
@@ -263,6 +263,35 @@ so that it works across screen sizes,
...
+- Depending on certain method names not being defined.
+- Depending on a feature from earlier specifications not being implemented as specified.
+
+When considering changes to widely used features, relative to current implementations and author expectations, the following questions should be considered:
+
+* Would removal of this feature or behavior have significant impact on existing content?
+* Does any of the dependent content occur on particularly high volume websites?
+* Is the dependent content genuinely intended for consumption, rather than occurring solely in test cases or examples?
+* Is the dependent content on the public web, rather than found solely on internal sites with a controlled user environment?
+* Does the dependent content currently work as intended in multiple popular user agents, rather than explicitly targeting only one particular user agent?
```
Comment (LeaVerou) [inline index.bs:287]:
```suggestion
When considering changes to widely used features, consider the impact of the change on existing content, weighing the prevalence and popularity of current usage, and interoperability of the old behavior. 
```

File: index.bs [hunk 33]
```diff
@@ -263,6 +263,35 @@ so that it works across screen sizes,
...
+
+The benefit of the proposed change should be weighed against the likely cost of breaking content, as measured by these criteria.
```
Comment (LeaVerou) [inline index.bs:289]:
```suggestion
```

(covered by suggestion above)

**Review by martinthomson — CHANGES_REQUESTED** [review]
Rather than nitpick this, how about:

# New features can add functionality to the Web

A new Web feature can add capabilities to the Web with little concern about the effect on existing content.  The Web includes many extension points that allow for additions.  This includes HTML elements and attributes, DOM interfaces and functions, CSS rules and selectors.

# Removing or changing capabilities needs particular care

An important goal of the Web is that it is relatively stable.
Modern user agents are generally capable of handling most older content.

This depends on not removing features and capabilities
without first gaining an understanding of the impact on existing content.
This might require an attempt to gain an understanding of
how features are used by existing content.

The obligation to understand existing usage also applies to any features that content relies upon.
This includes vendor-proprietary features and
behavior that might be considered implementation bugs.
Web features are not solely defined in specifications;
they are also defined by how content uses those features.

**Comment by LeaVerou** [comment]
Closing this as it was superseded by #469
